crates/sui-core/src/checkpoints/checkpoint_executor/data_ingestion_handler.rs

```diff
@@ -0,0 +1,152 @@
+            .into_iter()
+            .map(|kind| {
+                let (id, version) = kind.id_and_version();
+                ObjectKey(id, version)
+            })
+            .chain(
+                fx.modified_at_versions()
+                    .into_iter()
+                    .map(|(object_id, version)| ObjectKey(object_id, version)),
+            )
+            .collect::<HashSet<_>>()
+            .into_iter()
+            // Unwrapped-then-deleted objects are not stored in state before the tx, so we have nothing to fetch.
+            .filter(|key| !unwrapped_then_deleted_obj_ids.contains(&key.0))
+            .collect::<Vec<_>>();
+
+        let input_objects = authority_store
+            .multi_get_object_by_key(&input_object_keys)?
+            .into_iter()
+            .zip(&input_object_keys)
+            .map(|(object, object_key)| {
+                object.ok_or(SuiError::UserInputError {
+                    error: UserInputError::ObjectNotFound {
+                        object_id: object_key.0,
+                        version: Some(object_key.1),
+                    },
+                })
+            })
+            .collect::<SuiResult<Vec<_>>>()?;
+
+        let output_object_keys = fx
+            .all_changed_objects()
+            .into_iter()
+            .map(|(object_ref, _owner, _kind)| ObjectKey::from(object_ref))
+            .collect::<Vec<_>>();
+
+        let output_objects = authority_store
+            .multi_get_object_by_key(&output_object_keys)?
+            .into_iter()
+            .zip(&output_object_keys)
+            .map(|(object, object_key)| {
+                object.ok_or(SuiError::UserInputError {
+                    error: UserInputError::ObjectNotFound {
+                        object_id: object_key.0,
+                        version: Some(object_key.1),
+                    },
+                })
+            })
+            .collect::<SuiResult<Vec<_>>>()?;
+
+        let full_transaction = CheckpointTransaction {
+            transaction: tx.into(),
+            effects: fx,
+            events,
+            input_objects,
+            output_objects,
+        };
+        full_transactions.push(full_transaction);
+    }
+    let file_name = format!("{}.chk", checkpoint.sequence_number);
+    let checkpoint_data = CheckpointData {
+        checkpoint_summary: checkpoint.into(),
+        checkpoint_contents,
+        transactions: full_transactions,
+    };
+
+    std::fs::create_dir_all(&path).map_err(|err| {
+        SuiError::FileIOError(format!(
+            "failed to save full checkpoint content locally {:?}",
+            err
+        ))
+    })?;
+    let bytes =
+        bcs::to_bytes(&checkpoint_data).expect("failed to serialize full checkpoint content");
+    std::fs::write(path.join(file_name), bytes).map_err(|err| {
+        SuiError::FileIOError(format!(
+            "failed to save full checkpoint content locally {:?}",
+            err
+        ))
+    })?;
```

Comment:
You could avoid an allocation by opening the file and then using https://docs.rs/bcs/latest/bcs/fn.serialize_into.html to directly write the contents into the file.